mockbroker_test.go

```diff
@@ -47,80 +72,134 @@ func (b *mockBroker) Addr() string {
+
+	go func() {
+		<-b.closing
+		safeClose(b.t, b.listener)
+	}()
+
+	wg := &sync.WaitGroup{}
+	for i := 0; ; i++ {
+		if conn, err = b.listener.Accept(); err != nil {
+			goto done
+		}
+		wg.Add(1)
+		go b.handleRequests(conn, i, wg)
 	}
-	reqHeader := make([]byte, 4)
+done:
+	wg.Wait()
+	Logger.Printf("*** mockbroker/%d: listener closed, err=%v", b.BrokerID(), err)
+}
+
+func (b *mockBroker) handleRequests(conn net.Conn, idx int, wg *sync.WaitGroup) {
+	defer wg.Done()
+	defer func() {
+		_ = conn.Close()
+	}()
+	Logger.Printf("*** mockbroker/%d/%d: connection opened", b.BrokerID(), idx)
+	var err error
+
+	abort := make(chan none)
+	defer close(abort)
+	go func() {
+		select {
+		case <-b.closing:
+			_ = conn.Close()
+		case <-abort:
+		}
+	}()
+
 	resHeader := make([]byte, 8)
-	for expectation := range b.expectations {
-		_, err = io.ReadFull(conn, reqHeader)
+	for {
+		req, err := decodeRequest(conn)
 		if err != nil {
-			return b.serverError(err, conn)
-		}
-		body := make([]byte, binary.BigEndian.Uint32(reqHeader))
-		if len(body) < 10 {
-			return b.serverError(errors.New("Kafka request too short."), conn)
-		}
-		if _, err = io.ReadFull(conn, body); err != nil {
-			return b.serverError(err, conn)
+			Logger.Printf("*** mockbroker/%d/%d: invalid request: err=%v, %v", b.brokerID, idx, err, spew.Sdump(req))
+			b.serverError(err)
+			goto done
 		}
 
 		if b.latency > 0 {
 			time.Sleep(b.latency)
 		}
 
-		response, err := encode(expectation)
+		res := b.requestHandler()(req)
+		if _, ok := req.body.(*FetchRequest); !ok {
+			Logger.Printf("*** mockbroker/%d/%d: served %+v -> %+v", b.brokerID, idx, req, res)
+		}
+		var encodedRes []byte
+		encodedRes, err = encode(res)
 		if err != nil {
-			return false
+			b.serverError(err)
+			goto done
 		}
-		if len(response) == 0 {
+		if len(encodedRes) == 0 {
 			continue
 		}
 
-		binary.BigEndian.PutUint32(resHeader, uint32(len(response)+4))
-		binary.BigEndian.PutUint32(resHeader[4:], binary.BigEndian.Uint32(body[4:]))
+		binary.BigEndian.PutUint32(resHeader, uint32(len(encodedRes)+4))
+		binary.BigEndian.PutUint32(resHeader[4:], uint32(req.correlationID))
 		if _, err = conn.Write(resHeader); err != nil {
-			return b.serverError(err, conn)
+			b.serverError(err)
+			goto done
 		}
-		if _, err = conn.Write(response); err != nil {
-			return b.serverError(err, conn)
+		if _, err = conn.Write(encodedRes); err != nil {
+			b.serverError(err)
+			goto done
 		}
 	}
-	if err = conn.Close(); err != nil {
-		return b.serverError(err, nil)
-	}
-	if err = b.listener.Close(); err != nil {
-		b.t.Error(err)
-		return false
-	}
-	return true
+done:
+	Logger.Printf("*** mockbroker/%d/%d: connection closed, err=%v", b.BrokerID(), idx, err)
 }
 
-func (b *mockBroker) serverError(err error, conn net.Conn) bool {
-	b.t.Error(err)
-	if conn != nil {
-		if err := conn.Close(); err != nil {
-			b.t.Error(err)
+func (b *mockBroker) requestHandler() func(req *request) (res encoder) {
+	b.handlerLock.Lock()
```

Comment:
> In fact there is a use case for it. E.g. check out this particular consumer test rewritten using the updated broker. Without resetting the handler during the test the logic of the handler would be very complicated.

That test is still racy though, even with the locks... you'd need some other mechanism to actually make it deterministic. I think it would be better if `SetHandlerByMap` took a `map[string][]MockResponse` instead to avoid the need for updating the handler mid-test?

> And by the way what is wrong with using proper synchronization? :)

Nothing, unless it's unnecessary :)